// tsparticles-engine/browser/Core/Utils/InteractionManager.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _InteractionManager_engine, _InteractionManager_interactors;
export class InteractionManager {
    /**
     * The constructor of the interaction manager
     * @param engine the parent engine
     * @param container the parent container
     */
    constructor(engine, container) {
        this.container = container;
        /**
         * The engine used for registering the interactions managers
         * @private
         */
        _InteractionManager_engine.set(this, void 0);
        /**
         * The interactors that are used for initialization
         * @private
         */
        _InteractionManager_interactors.set(this, void 0);
        __classPrivateFieldSet(this, _InteractionManager_engine, engine, "f");
        __classPrivateFieldSet(this, _InteractionManager_interactors, __classPrivateFieldGet(this, _InteractionManager_engine, "f").plugins.getInteractors(this.container, true), "f");
        this.externalInteractors = [];
        this.particleInteractors = [];
    }
    /**
     * Iterates through the external interactivity manager and call the interact method, if they are enabled
     * @param delta this variable contains the delta between the current frame and the previous frame
     */
    async externalInteract(delta) {
        for (const interactor of this.externalInteractors) {
            if (interactor.isEnabled()) {
                await interactor.interact(delta);
            }
        }
    }
    handleClickMode(mode) {
        for (const interactor of this.externalInteractors) {
            if (interactor.handleClickMode) {
                interactor.handleClickMode(mode);
            }
        }
    }
    /**
     * Initializes the interaction manager, loading all the engine registered managers into the container
     */
    init() {
        this.externalInteractors = [];
        this.particleInteractors = [];
        for (const interactor of __classPrivateFieldGet(this, _InteractionManager_interactors, "f")) {
            switch (interactor.type) {
                case 0 /* InteractorType.External */:
                    this.externalInteractors.push(interactor);
                    break;
                case 1 /* InteractorType.Particles */:
                    this.particleInteractors.push(interactor);
                    break;
            }
            interactor.init();
        }
    }
    /**
     * Iterates through the particles interactions manager and call the interact method, if they are enabled
     * @param particle the particle responsible for the current interaction
     * @param delta this variable contains the delta between the current frame and the previous frame
     */
    async particlesInteract(particle, delta) {
        for (const interactor of this.externalInteractors) {
            interactor.clear(particle);
        }
        /* interaction auto between particles */
        for (const interactor of this.particleInteractors) {
            if (interactor.isEnabled(particle)) {
                await interactor.interact(particle, delta);
            }
        }
    }
    /**
     * Iterates through the external interactivity manager and call the interact method, if they are enabled
     * @param particle the particle to reset
     */
    async reset(particle) {
        for (const interactor of this.externalInteractors) {
            if (interactor.isEnabled()) {
                await interactor.reset(particle);
            }
        }
        for (const interactor of this.particleInteractors) {
            if (interactor.isEnabled(particle)) {
                await interactor.reset(particle);
            }
        }
    }
}
_InteractionManager_engine = new WeakMap(), _InteractionManager_interactors = new WeakMap();
